src/mscorlib/src/System/StubHelpers.cs

```diff
@@ -125,6 +125,112 @@ static internal void ClearNative(IntPtr pNative)
         }
     }  // class CSTRMarshaler
 
+    [ReliabilityContract(Consistency.WillNotCorruptState, Cer.MayFail)]
+    internal static class UTF8Marshaler
+    {
+        const int UTF8_CHAR_SIZE = 3;
+        static internal unsafe IntPtr ConvertToNative(int flags, string strManaged, IntPtr pNativeBuffer)
+        {
+            if (null == strManaged)
+            {
+                return IntPtr.Zero;
+            }
+            StubHelpers.CheckStringLength(strManaged.Length);
+
+            int nb;
+            byte* pbNativeBuffer = (byte*)pNativeBuffer;
+
+            // If we are marshaling into a stack buffer allocated by the ILStub
+            // we will use a "1-pass" mode where we convert the string directly into the unmanaged buffer.   
+            // else we will allocate the precise native heap memory.          
+            if (pbNativeBuffer != null)
+            {
+                // this is the number of bytes allocated by the ILStub.
+                nb = (strManaged.Length + 1) * UTF8_CHAR_SIZE;
+
+                // +1 for the '\0' that we add
+                nb += 1;
```

Comment:
What you saying is correct , but if i take out the +1 then it will obscure the fact that we null terminate the string and the bytes allocated by ILStub is actually number of bytes  + 1 . Anyway i will take it out.